### PR TITLE
Over The Air Flickering Evasion Attack

### LINKED ISSUE #881: Implement Over-the-Air Adversarial Flickering Attacks against Video Recognition Networks
Paper link: https://arxiv.org/abs/2002.05123

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request continues from #1077 to add final changes to integrate the new attack

Fixes #881

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        "start_frame_index",
+        "num_frames",
+        "round_samples",
+    ]
+
+    _estimator_requirements = (BaseEstimator, ClassifierMixin, LossGradientsMixin)
+
+    def __init__(
+        self,
+        classifier: "CLASSIFIER_LOSS_GRADIENTS_TYPE",
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:70]: I think this should only be a type of PyTorchClassifier

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+        self.end_frame_index = (
+            self.start_frame_index + self.num_frames if self.num_frames is not None else self.num_frames
+        )
+        self._check_params()
+
+    def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> np.ndarray:
+        """
+        Generate adversarial examples.
+
+        :param x: Original input samples.
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:118]: It is worth to specify the shape of `x` here because `x` now is not an image.

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+                :, self.start_frame_index : self.end_frame_index, :, :, :
+            ]
+            grad = full_grad
+
+        return grad
+
+    def _get_loss_gradients(self, x: "torch.Tensor", y: "torch.Tensor", perturbation: "torch.Tensor") -> "torch.Tensor":
+        """
+        Compute custom, framework-specific, regularized, loss gradients.
+
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:230]:
```suggestion
```

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+            full_grad[:, self.start_frame_index : self.end_frame_index, :, :, :] = grad[
+                :, self.start_frame_index : self.end_frame_index, :, :, :
+            ]
+            grad = full_grad
+
+        return grad
+
+    def _get_loss_gradients(self, x: "torch.Tensor", y: "torch.Tensor", perturbation: "torch.Tensor") -> "torch.Tensor":
+        """
+        Compute custom, framework-specific, regularized, loss gradients.
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:229]:
```suggestion
        Compute custom, framework-specific, regularized loss gradients.
```

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+
+        """
+        import torch  # lgtm [py/repeated-import]
+
+        softmax = torch.nn.Softmax(dim=1).to(self.estimator.device)
+
+        grads_batch = []
+
+        for i in range(x.shape[0]):
+
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:239]:
```suggestion
```

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+    def _compute_torch(
+        self,
+        x_adv: "torch.Tensor",
+        x: "torch.Tensor",
+        y: "torch.Tensor",
+        eps_step: float,
+    ) -> "torch.Tensor":
+        """
+        Compute adversarial examples for one iteration.
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:182]:
```suggestion
        Compute adversarial examples for one iteration.
        
```

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+            x = torch.clamp(x, self.estimator.clip_values[0], self.estimator.clip_values[1])
+
+        if self.round_samples != 0.0:
+            x = torch.round(x / self.round_samples) * self.round_samples
+
+        return x
+
+    def _apply_perturbation(self, x_adv: "torch.Tensor", grad: "torch.Tensor", eps_step: float) -> "torch.Tensor":
+        """
+        Apply perturbation on examples.
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:311]:
```suggestion
        Apply perturbation on examples.
        
```

File: tests/attacks/evasion/test_over_the_air.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        super(Model, self).__init__()
+        self.fc = torch.nn.Linear(12 * 299 * 299 * 3, 101)
+
+    def forward(self, x):
+        x = x.view(-1, 12 * 299 * 299 * 3)
+        logit_output = self.fc(x)
+        return logit_output.view(-1, 101)
+
+
+@pytest.mark.only_with_platform("pytorch")
```
Comment (minhitbk) [inline tests/attacks/evasion/test_over_the_air.py:46]: This is deprecated, please turn to use @pytest.mark.skip_framework

File: tests/attacks/evasion/test_over_the_air.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            x=torch.from_numpy(x_train), y=torch.from_numpy(y_train), perturbation=torch.zeros(x_train.shape)
+        )
+
+        assert gradients.shape == (2, 12, 1, 1, 3)
+
+    except ARTTestException as e:
+        art_warning(e)
+
+
+@pytest.mark.only_with_platform("pytorch")
```
Comment (minhitbk) [inline tests/attacks/evasion/test_over_the_air.py:70]: This is deprecated, please turn to use @pytest.mark.skip_framework

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+            raise ValueError("The argument `batch_size` must be positive of type int.")
+
+        if not isinstance(self.start_frame_index, int) or self.start_frame_index < 0:
+            raise ValueError("The argument `start_frame_index` must be 0 or positive of type int.")
+
+        if self.num_frames is not None and (not isinstance(self.num_frames, int) or self.num_frames <= 0):
+            raise ValueError("The argument `num_frames` must be positive of type int.")
+
+        if not isinstance(self.round_samples, (int, float)) or self.round_samples < 0.0:
+            raise ValueError("The argument `round_samples` must be 0 .0 or positive of type int or float.")
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:360]:
```suggestion
            raise ValueError("The argument `round_samples` must be 0.0 or positive of type int or float.")
```

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
+                  (nb_samples,).
+        :return: Adversarial examples.
+        """
+        import torch  # lgtm [py/repeated-import]
+
+        y = check_and_transform_label_format(y, self.estimator.nb_classes)
+        if y is None:
+            logger.info("Using model predictions as true labels.")
+            y = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:128]: I think this version of attack only supports targeted attacks, so `y` must be provided by users, otherwise we need to raise an error. The labels returned by model prediction are only used when untargeted attacks are supported.

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+        y = y.to(self.estimator.device)
+        x_adv = torch.clone(x)
+
+        for _ in range(self.max_iter):
+            x_adv = self._compute_torch(
+                x_adv,
+                x,
+                y,
+                self.eps_step,
+            )
```
Comment (minhitbk) [inline art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:170]:
```suggestion
            x_adv = self._compute_torch(x_adv, x, y, self.eps_step)
```

**Review by minhitbk — CHANGES_REQUESTED** [review]
I think overall the implementation is good. I only have a few comments, most of them are minor. There is only one that I concern, which is the use of model prediction as targeted labels when `y` is not provided by users. I think this use is only for when we have untargeted attack supported, where model prediction is used as `target` for the untargeted attack, i.e. to make an adversarial example produce something different from the `target`.

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py (hunk 1 above)
Comment (beat-buesser) [reply art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:70]: I agree.

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py (hunk 2 above)
Comment (beat-buesser) [reply art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:118]: I agree.

File: tests/attacks/evasion/test_over_the_air.py (hunk 8 above)
Comment (beat-buesser) [reply tests/attacks/evasion/test_over_the_air.py:46]: I agree.

File: tests/attacks/evasion/test_over_the_air.py (hunk 9 above)
Comment (beat-buesser) [reply tests/attacks/evasion/test_over_the_air.py:70]: I agree.

File: art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py (hunk 12 above)
Comment (beat-buesser) [reply art/attacks/evasion/over_the_air_flickering/over_the_air_flickering_pytorch.py:170]: I didn't apply this change, because the existing version is preferred by `black==20.8b1`.

**Comment by beat-buesser** [comment]
@minhitbk Thank you very much for your great review! I'm merging according to our offline discussion.
